### DISCUSSION TITLE
Reorder Nodes in SLEAP Skeleton

### DISCUSSION THREAD

**Discussion by ghost** [talmolab/sleap#1199] [Help!] [answered]
I would like to know if there is an easy way to rearrange the order of the nodes in my skeleton? As of now, node 1 is nose, node 2 is L ear, node 3 is R ear, and node 4 is tail. I'd like to rearrange it so that node 1 is L ear, node 2 is R ear, node 3 is nose, and node 4 is tail.

Thanks much!

**Comment by roomrys** [comment]
Hi @umaisk,

To help me better answer, is there something specific you would like to accomplish by rearranging the order of the nodes in your skeleton?

Thanks,
Liezl

**Reply by ghost** [reply]
I am trying to feed the SLEAP pose-estimation data into SimBA (supervised classifier) to pull out behavioral syllables. The nodes have to be ordered in a specific way for me to import the SLEAP data into SimBA.

**Reply by roomrys** [reply]
Hi @umaisk,

There is a way to rearrange the nodes by recreating [the underlying graph](https://networkx.org/documentation/stable/reference/classes/digraph.html). I can work on getting you a stand-alone solution, but for a long-term solution, it might be best integrated in the SLP -> Simba adaptor, i.e [creating a request on the Simba repo](https://github.com/sgoldenlab/simba/issues/new/choose). Perhaps our stand-alone code can help implement something permanent there.

Tagging @sronilsson in case they have any insight on the node-ordering requirement.

Thanks,
Liezl

**Reply by sronilsson** [reply]
Hi @roomrys - there are some hard-coded [featurizers in simba](https://github.com/sgoldenlab/simba/blob/master/simba/features_scripts/feature_extractor_16bp.py). These are used when users create projects with [pre-defined pose configurations](https://github.com/sgoldenlab/simba/blob/master/simba/pose_configurations/schematics/Picture3.png), which require specific node orders. 

There is a method to [re-order pose data](https://github.com/sgoldenlab/simba/blob/master/docs/Tutorial_tools.md#re-organize-tracking-data)  in cases where users wants to use a hard-coded featurizer, and have pose data for the specific body-parts, but nodes are in the "wrong" order. Unfortunately this method doesn't support sleap h5/slp and I am not sure how to go about it for sleap h5/slp files.

**Comment by roomrys** [comment, marked as answer]
Apologies for the delay! I wrote a MVP script to handle the reordering (see below). I am hesitant to add this to SLEAP since it is functionality that SLEAP itself would never call on, but if it is something that Simba (or other softwares) would use, then I could add an API for it?
```python
from typing import List

import networkx as nx

import sleap
from sleap.skeleton import EdgeType, Skeleton
from sleap.gui.app import main as sleap_label


def reorder_graph(graph: nx.MultiDiGraph, new_order: List[int]):
    """Reorder the nodes in the graph in order specified by `new_order`.

    Args:
        graph: The graph to reorder.
        new_order: List of integers specifying the new order of the nodes. Must be the
            same length as the number of nodes in the graph. For example, if the
            original graph has 4 nodes, and you want to reverse the order, then
            `new_order` should be [3, 2, 1, 0]. Node 0 -> 3, node 1-> 2, node 2 -> 1,
            node 3 -> 0.

    Returns:
        New graph with the nodes reordered.
    """

    # Create new graph
    new_graph = nx.MultiDiGraph(name=graph.graph["name"], num_edges_inserted=0)

    # Add nodes to new graph in new order
    reordered_nodes = [node for _, node in sorted(zip(new_order, graph._node.items()))]
    for node, d in reordered_nodes:
        new_graph.add_node(node)

    # Add edges exactly the same as the original graph
    for source_node, destination_node in graph.edges():
        new_graph.add_edge(
            source_node,
            destination_node,
            type=EdgeType.BODY,
            edge_insert_idx=new_graph.graph["num_edges_inserted"],
        )
        new_graph.graph["num_edges_inserted"] = (
            new_graph.graph["num_edges_inserted"] + 1
        )
    return new_graph


def main(ds, new_order: List[int]):
    """Reorder the nodes in the graph in order specified by `new_order`.

    Args:
        ds: Path to the dataset.
        new_order: List of integers specifying the new order of the nodes. Must be the
            same length as the number of nodes in the graph. For example, if the
            original graph has 4 nodes, and you want to reverse the order, then
            `new_order` should be [3, 2, 1, 0]. Node 0 -> 3, node 1-> 2, node 2 -> 1,
            node 3 -> 0.

    Returns:
        Path to the new dataset.
    """

    # Load the graph from the file.
    labels = sleap.load_file(ds)
    skeleton: Skeleton = labels.skeleton
    graph: nx.MultiDiGraph = skeleton._graph

    # Reverse the graph.
    new_graph = reorder_graph(graph, new_order)

    # Replace the graph in the Skeleton object.
    skeleton._graph = new_graph

    # Save the labels
    new_ds = f"{ds}_reordered.slp"
    labels.save(new_ds)

    return new_ds


if __name__ == "__main__":
    # Load your dataset here.
    ds = r"path\to\your\dataset.slp"

    # Do the reordering.
    new_order = [2, 0, 1, 3]  # Node 0 -> 2, node 1-> 0, node 2 -> 1, node 3 -> 3
    new_ds = main(ds, new_order)

    # Open the new dataset in SLEAP to double check everything looks good!
    sleap_label([new_ds])

```
